src/log-store/src/raft_engine/backend.rs

```diff
@@ -0,0 +1,548 @@
+                succeeded = false;
+                break;
+            }
+        }
+
+        let mut batch = LogBatch::default();
+        let do_txn = |txn_op| match txn_op {
+            TxnOp::Put(key, value) => {
+                batch
+                    .put(SYSTEM_NAMESPACE, key.clone(), value)
+                    .context(RaftEngineSnafu)?;
+                Ok(TxnOpResponse::ResponsePut(PutResponse { prev_kv: None }))
+            }
+
+            TxnOp::Get(key) => {
+                let value = engine_get(&engine, &key)?.map(|kv| kv.value);
+                let kvs = value
+                    .into_iter()
+                    .map(|value| KeyValue {
+                        key: key.clone(),
+                        value,
+                    })
+                    .collect();
+                Ok(TxnOpResponse::ResponseGet(RangeResponse {
+                    kvs,
+                    more: false,
+                }))
+            }
+
+            TxnOp::Delete(key) => {
+                let prev_value = engine_get(&engine, &key)?;
+                batch.delete(SYSTEM_NAMESPACE, key);
+
+                let prev_kvs: Vec<_> = prev_value.into_iter().collect();
+                let deleted = prev_kvs.len() as i64;
+
+                Ok(TxnOpResponse::ResponseDelete(DeleteRangeResponse {
+                    deleted,
+                    prev_kvs,
+                }))
+            }
+        };
+
+        let responses = if succeeded { success } else { failure }
+            .into_iter()
+            .map(do_txn)
+            .collect::<error::Result<_>>()?;
+
+        engine.write(&mut batch, false).context(RaftEngineSnafu)?;
+
+        Ok(TxnResponse {
+            succeeded,
+            responses,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl KvBackend for RaftEngineBackend {
+    fn name(&self) -> &str {
+        "RaftEngineBackend"
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    async fn range(&self, req: RangeRequest) -> Result<RangeResponse, Self::Error> {
+        let mut res = vec![];
+        self.engine
+            .read()
+            .unwrap()
+            .scan_raw_messages(
+                SYSTEM_NAMESPACE,
+                Some(&req.key),
+                Some(&req.range_end),
+                false,
+                |key, value| {
+                    res.push(KeyValue {
+                        key: key.to_vec(),
+                        value: value.to_vec(),
+                    });
+                    true
+                },
+            )
+            .context(RaftEngineSnafu)?;
+        Ok(RangeResponse {
+            kvs: res,
+            more: false,
+        })
+    }
+
+    async fn put(&self, req: PutRequest) -> Result<PutResponse, Self::Error> {
+        let PutRequest {
+            key,
+            value,
+            prev_kv,
+        } = req;
+
+        let mut prev = None;
+        let engine = self.engine.read().unwrap();
```

Comment:
> Maybe obtain write lock?
> 
> Also I've found there're other places in this file that might instead acquire write lock, please check them.

`Engine::write` guarantees that the logbatch is atomatically written. Thus we won't have write-write conflict. The read lock here is just for prevent race condition between `KvBackend::put` and `KvBackend::txn` to avoid txn being affected by writes.